typed-racket-lib/typed-racket/base-env/prims-contract.rkt

```diff
@@ -250,9 +251,17 @@
 ;;  make-predicate
 ;;  cast
 
-;; Helper to construct syntax for contract definitions
+;; Helpers to construct syntax for contract definitions
+;; make-contract-def-rhs : Type-Stx Boolean Boolean -> Syntax
 (define (make-contract-def-rhs type flat? maker?)
-  (contract-def-property #'#f `#s(contract-def ,type ,flat? ,maker? untyped)))
+  (define contract-def `#s(contract-def ,type ,flat? ,maker? untyped))
+  (contract-def-property #'#f (λ () contract-def)))
+
+;; make-contract-def-rhs/from-typed : Id Boolean Boolean -> Syntax
+(define (make-contract-def-rhs/from-typed id flat? maker?)
+  (contract-def-property
+   #'#f
+   (λ () `#s(contract-def ,(or (cast-table-ref id) #'Any) ,flat? ,maker? typed))))
```

Comment:
It might be because of the sandboxed evaluators, because this also fails in the "thunk called too early" version:
```racket
#lang racket
(require racket/sandbox)
(define ev (call-with-trusted-sandbox-configuration
             (λ () (make-module-evaluator "#lang typed/racket"))))
(ev '(cast 3 Integer))
```